krm-functions/lib/kptfile/v1/condition.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2023 Nephio.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GetConditionType returns a string based on the KRM object Reference
+// It acts on APIVersion, Kind and Name. if these parameters dont exist
+// the string does not contain these parameters
+func GetConditionType(o *corev1.ObjectReference) string {
+	var sb strings.Builder
+	sb.Reset()
+	if o.APIVersion != "" {
+		gv, err := schema.ParseGroupVersion(o.APIVersion)
+		if err == nil {
+			sb.WriteString(gv.String())
+		}
+	}
+	if o.Kind != "" {
+		if sb.String() != "" {
+			sb.WriteString(".")
+		}
+		sb.WriteString(o.Kind)
+	}
+	if o.Name != "" {
+		if sb.String() != "" {
+			sb.WriteString(".")
+		}
+		sb.WriteString(o.Name)
+	}
+	return sb.String()
+}
+
+// GetGVKNFromConditionType return a KRM ObjectReference from a string
+// It expects an APIVersion with a / as a.b/c and a kind + name
+// if not it retruns an empty ObjectReference
+func GetGVKNFromConditionType(ct string) (o *corev1.ObjectReference) {
```

Comment:
Do we need this reverse mapping? If we are using it in code, it locks us in pretty tightly to that naming schema. I can see it might be needed, for example, if we need to clean up stale conditions. Hmm.

Reply:
GetConditionType does corev1.ObjectReference -> string
GetGVKNFromConditionType does string -> corev1.ObjectReference

Maybe the naming should be more aligned?

Reply:
I understood that, I think the name is fine. My question was more about whether we need it. My philosophy is generally not to add any code until we need it. One less thing to test and maintain, if it's not being used.

Reply:
Here is how it is used in the fn-runtime.

// collect all conditions
		for _, c := range kf.GetConditions() {
			objRef := *kptfilelibv1.GetGVKNFromConditionType(c.Type)
			if len(r.cfg.Owns) == 0 {
				// collect all conditions for a wildcard
				r.inv.AddCondition(Wildcard, &objRef, &c)
			} else {
				for ref := range r.cfg.Owns {
					if strings.Contains(c.Type, kptfilelibv1.GetConditionType(&ref)) {
						r.inv.AddCondition(objRef.Name, &objRef, &c)
					}
				}
			}
		}
		
1. We track the conditions from the kptfile. E.g. a user can delete them and we want to ensure we are always aligned in an idempotent way to get back to the proper state.
2. we track the condition to tell what work is to be done and we use corev1.Objectoreference to store the data in a map to make lookup easier.

Reply:
Ok, thanks.